framework/db/BaseActiveRecord.php

```diff
@@ -1235,8 +1235,13 @@ public function link($name, $model, $extraColumns = [])
             $this->_related[$name] = $model;
         } elseif (isset($this->_related[$name])) {
             if ($relation->indexBy !== null) {
-                $indexBy = $relation->indexBy;
-                $this->_related[$name][$model->$indexBy] = $model;
+                if ($relation->indexBy instanceof \Closure) {
+                    $index = call_user_func($relation->indexBy, $model);
+                } else {
+                    $indexBy = $relation->indexBy;
+                    $index = $model->$indexBy;
```

Comment:
I return back to this:
```php
                    $indexBy = $relation->indexBy;
                    $index = $model->$indexBy;
```
because when test in with PHP 5.5 there was error.